### PR TITLE
Implementing a Recordable interface for logs

### PR DESCRIPTION
**Pull request by kxyr**
This PR implements a recordable for logs, resolving issue # 412. This idea essentially delays creation of a log record until the exporter is called (instead of creating it in the Logging API as it does currently). 

This PR makes the following changes:

* Adds a Recordable interface for logs
* Add a default implementation of the Recordable, a `LogRecord` class (similar to `SpanData` for traces)
* Updates the methods and signatures of the the simple processor, processor interface, and exporter interface, including adding a MakeRecordable() method
* Refactors all the Logging API, SDK, and Simple Processor tests

API/SDK changes: 
* Removes the `LogRecord` struct from the API 
* Removed the `Severity::kDefault` enum value (originally set as kInfo), and added a `Severity::kInvalid`= 0. (Setting the default Severity to kInfo has been moved to the SDK instead of the API)
* Gets the TraceId, SpanId, TraceFlags, and current Timestamp in the SDK and injects into the LogRecord if it is not user-specified. 

Other minor changes: 
* Renames trace_flag to trace_flags in LogRecord 
* Stylistic change: Renames the log() methods to Log() to follow this repository's naming convention, and calling this->Log() instead of simple Log() for clarity

The call path currently looks like this: 
* `api::Logger::Log()` calls -> `sdk::Logger::Log()` -> creates a recordable by calling -> `Processor::MakeRecordable()` which calls - > `Exporter::MakeRecordable()` -> (which can return a default `LogRecord` Recordable, similar to `SpanData` for traces) -> SDK populates recordable it receives, then  calls -> `Processor::OnReceive(recordable)` calls -> `Exporter::Export(span of recordables)` to export. 

To keep this Recordable PR short and focused, there are only a very limited number of Logging API methods. In another PR, more overloads will be added to make the API more ergonomic and flexible to use. 

Note on code coverage: Additional tests for this Recordable interface (in the processor, exporters, SDK) will be added, but will be covered by another future PR. The current unit tests do include calls of `MakeRecordable()` which ensures this is working properly for common cases. 

This PR introduces changes that affect a few other open PRs. It would be better having this PR merged before PR # 422, # 430, # 434 , and # 435, so those can be rebased and updated accordingly. 

cc @MarkSeufert @alolita

### CODE REVIEW

**Comment by kxyr** [comment]
Failing CI seems to be unrelated to this PR - (Traces CircularBufferTest) - it usually sorts itself out after a period of time.

File: sdk/include/opentelemetry/sdk/logs/logger.h [hunk 1]
```diff
@@ -45,9 +44,26 @@ class Logger final : public opentelemetry::logs::Logger
 
   /**
    * Writes a log record into the processor.
-   * @param record The record to write into the processor.
-   */
-  void log(const opentelemetry::logs::LogRecord &record) noexcept override;
+   * @param timestamp the timestamp the log record was created.
+   * @param name the name of the log event.
```
Comment (ranton256) [inline sdk/include/opentelemetry/sdk/logs/logger.h]: These are not in matching order with function signature.

File: sdk/include/opentelemetry/sdk/logs/logger.h [hunk 2]
```diff
@@ -45,9 +44,26 @@ class Logger final : public opentelemetry::logs::Logger
...
+   * @param severity the severity level of the log event.
+   * @param message the string message of the log (perhaps support std::fmt or fmt-lib format).
+   * @param record the log record (object type LogRecord) that is logged.
```
Comment (ranton256) [inline sdk/include/opentelemetry/sdk/logs/logger.h]: record param is gone

Comment (ranton256) [inline sdk/include/opentelemetry/sdk/logs/logger.h]: The body and resource params are missing from comment

File: api/include/opentelemetry/logs/logger.h [hunk 3]
```diff
@@ -43,81 +61,103 @@ class Logger
...
+           nostd::string_view body,
+           std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>> resource,
+           std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>> attributes,
+           trace::TraceId trace_id,
+           trace::SpanId span_id,
+           trace::TraceFlags trace_flags) noexcept
   {
-    // TODO: set default timestamp later (not in API)
-    log(severity, message, core::SystemTimestamp(std::chrono::system_clock::now()));
+    return this->Log(timestamp, severity, name, body,
```
Comment (ranton256) [inline api/include/opentelemetry/logs/logger.h:126]: why the explicit "this->"?

File: sdk/include/opentelemetry/sdk/logs/logger_provider.h [hunk 4]
```diff
@@ -19,6 +19,7 @@
 #include <memory>
 #include <mutex>
 #include <string>
+#include <unordered_map>
```
Comment (ranton256) [inline sdk/include/opentelemetry/sdk/logs/logger_provider.h:22]: This is only line in file's diff so not how this include for unordered_map could be used?

**Review by ranton256 — COMMENTED** [review]
Looks good other than minor stuff. I like the Recordable interface, should be a cleaner abstraction to leave storage specifics more open.

File: api/include/opentelemetry/logs/logger.h (hunk 3 above)
Comment (kxyr) [reply api/include/opentelemetry/logs/logger.h:126]: This is to follow the traces' [StartSpan()](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/tracer.h#L52) convention :)

File: sdk/include/opentelemetry/sdk/logs/logger_provider.h (hunk 4 above)
Comment (kxyr) [reply sdk/include/opentelemetry/sdk/logs/logger_provider.h:22]: I had actually not tested compiling it before :) The unordered_map is need for the list of Loggers stored by the loggerProvider

**Pull request closed without merging**